Create new app from template

meisenzahl opened the pull request:
To further develop **Code** into a user-friendly development environment, I thought it would be a good idea to add an option to create an app from a template. This PR is a functional prototype for it.

Could be interesting especially in conjunction with https://github.com/elementary/code/pull/940.

It adds an option to create a new app.

![Bildschirmfoto von 2021-04-23 18 21 45](https://user-images.githubusercontent.com/10796736/115901311-4ae16300-a461-11eb-9a55-117c84fe1249.png)

Selecting this option opens a dialog asking for information about the app and the developer. When all the information is entered, the template folder `data/templates/app` is copied to the desired location in the background. In the filenames and in the files template tags like `{{ app_name }}` are used. These are then replaced with the information from the dialog.

![Bildschirmfoto von 2021-04-23 18 22 06](https://user-images.githubusercontent.com/10796736/115901322-4f0d8080-a461-11eb-822b-94efd5f67228.png)

In case of an error, a warning is displayed.

![Bildschirmfoto von 2021-04-23 18 22 38](https://user-images.githubusercontent.com/10796736/115901337-53399e00-a461-11eb-807b-f58952415c13.png)

The created app supports the following features:

- `Hdy.Window`
- `Hdy.HeaderBar`
- `Hdy.WindowHandle`
- settings to save/restore window position
- supports dark preference
- translations
- Flatpak manifest
- GitHub CI configuration

![Bildschirmfoto von 2021-04-23 18 24 38](https://user-images.githubusercontent.com/10796736/115901349-5765bb80-a461-11eb-8188-54be65cbadbd.png)

I am looking forward to ideas how to develop the feature further.

meisenzahl commented on .github/workflows/main.yml:
```diff
@@ -17,8 +17,6 @@ jobs:
         apt update
         apt install -y exuberant-ctags libeditorconfig-dev libgail-3-dev libgee-0.8-dev libgit2-glib-1.0-dev libgranite-dev libgtk-3-dev libgtksourceview-4-dev libgtkspell3-3-dev libhandy-1-dev libpeas-dev libsoup2.4-dev libvala-dev libvte-2.91-dev meson valac
     - name: Build
-      env:
```
This is removed because `install_subdir` in `data/meson.build` only works with absolute paths:

```
ValueError: dst_dir must be absolute, got out/usr/local/share/io.elementary.code/templates
```

Reference: https://github.com/elementary/code/runs/2421607609

danrabbit commented on data/templates/app/data/com.github.{{ github_username }}.{{ github_repository }}.appdata.xml.in:
```diff
@@ -0,0 +1,16 @@
+  <id>com.github.{{ github_username }}.{{ github_repository }}</id>
+  <metadata_license>CC0</metadata_license>
+  <name>{{ app_name }}</name>
+  <summary>A Catchy Tagline</summary>
+  <description>
+    <p>A quick summary of your app's main selling points and features. Just a couple sentences per paragraph is best.</p>
+  </description>
+  <custom>
+    <value key="x-appcenter-color-primary">#603461</value>
+    <value key="x-appcenter-color-primary-text">rgb(255, 255, 255)</value>
+    <value key="x-appcenter-suggested-price">5</value>
```
This kind of stuff I feel like we should either fill from the dialog at the beginning or leave blank. It wouldn't be good to have apps submitted with this dummy data pre-filled

meisenzahl replied:
I would suggest that we leave this blank and I create a follow up PR to add more fields to the form.

danrabbit commented on data/templates/app/data/com.github.{{ github_username }}.{{ github_repository }}.desktop.in:
```diff
@@ -0,0 +1,10 @@
+[Desktop Entry]
+Name={{ app_name }}
+GenericName={{ app_name }}
+Comment=Proves that we can use Vala and Gtk
```
`GenericName` shouldn't be the same as `Name` and `Comment` here should probably match the `summary` tag in appdata. Again, I don't think we should have dummy data

danrabbit commented on data/templates/app/data/com.github.{{ github_username }}.{{ github_repository }}.desktop.in:
```diff
@@ -0,0 +1,10 @@
+Name={{ app_name }}
+GenericName={{ app_name }}
+Comment=Proves that we can use Vala and Gtk
+Categories=Utility;Education;
+Exec=com.github.{{ github_username }}.{{ github_repository }}
+Icon=com.github.{{ github_username }}.{{ github_repository }}
+Terminal=false
+Type=Application
+Keywords=Hello;World;Example;
```
Probably especially bad to have keyword and category dummy data

danrabbit commented on data/templates/app/data/meson.build:
```diff
@@ -0,0 +1,30 @@
+icon_sizes = ['16', '24', '32', '48', '64', '128']
+
+foreach i : icon_sizes
+    install_data(
+        join_paths('icons', i, meson.project_name() + '.svg'),
+        install_dir: join_paths(get_option('datadir'), 'icons', 'hicolor', i + 'x' + i, 'apps')
+    )
+endforeach
```
This should also install to hidpi paths

danrabbit commented on data/templates/app/po/extra/meson.build:
```diff
@@ -0,0 +1,8 @@
+i18n.gettext('extra',
+  args: [
+    '--directory=' + meson.source_root(),
+    '--from-code=UTF-8'
```
I'm pretty sure you don't need to set `--from-code` with the `glib` preset

danrabbit commented on data/templates/app/po/meson.build:
```diff
@@ -0,0 +1,8 @@
+i18n.gettext(meson.project_name(),
```
I think this should also use the glib preset

danrabbit commented on data/templates/app/src/Application.vala:
```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) {{ current_year }} {{ your_name }} (https://github.com/{{ github_username }})
```
You don't need the `(c)` here
```suggestion
 * Copyright {{ current_year }} {{ your_name }} (https://github.com/{{ github_username }})
```

danrabbit commented on data/templates/app/src/Application.vala:
```diff
@@ -0,0 +1,74 @@
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
```
The dialog says it's using GPL 3, but this is a GPL 2 license

danrabbit commented on data/templates/app/src/MainWindow.vala:
```diff
@@ -0,0 +1,93 @@
+ * Authored by: {{ your_name }} <{{ your_email }}>
+ */
+
+public class MainWindow : Hdy.Window {
```
This should probably be an `ApplicationWindow` since that adds the build in stuff for actions
```suggestion
public class MainWindow : Hdy.ApplicationWindow {
```

danrabbit commented on data/templates/app/src/MainWindow.vala:
```diff
@@ -0,0 +1,93 @@
+    private Hdy.HeaderBar headerbar;
+    private uint configure_id;
+
+    public MyApp app { get; construct; }
```
Since this is a Gtk.Window subclass it already has a property for a `Gtk.Application` so this isn't necessary

danrabbit commented on data/templates/app/src/MainWindow.vala:
```diff
@@ -0,0 +1,93 @@
+ */
+
+public class MainWindow : Hdy.Window {
+    private Hdy.HeaderBar headerbar;
```
`headerbar` isn't accessed outside of `construct {}` so it doesn't need to be scoped to the whole class

danrabbit commented on data/templates/app/src/MainWindow.vala:
```diff
@@ -0,0 +1,93 @@
+
+    public MyApp app { get; construct; }
+
+    public MainWindow (MyApp app) {
+        Object (
+            app: app
+        );
+    }
+
+    construct {
+        Hdy.init ();
+
+        set_application (app);
```
Instead of this, you can set `application` directly in the `Object ()` call

danrabbit commented on data/templates/app/src/MainWindow.vala:
```diff
@@ -0,0 +1,93 @@
+        set_application (app);
+
+        headerbar = new Hdy.HeaderBar () {
+            decoration_layout = "close:",
```
Probably shouldn't hardcode the button layout like this when it's not necessary, and especially because this removes the ability to maximize

danrabbit commented on data/templates/app/src/MainWindow.vala:
```diff
@@ -0,0 +1,93 @@
+            title = "{{ app_name }}"
+        };
+
+        var content = new Gtk.Grid () {
+            hexpand = true,
+            vexpand = true,
+            halign = Gtk.Align.CENTER,
+            valign = Gtk.Align.CENTER
+        };
+
+        var label = new Gtk.Label (_("Hello Again World!"));
+        label.get_style_context ().add_class (Granite.STYLE_CLASS_H1_LABEL);
+        content.add (label);
```
This is all like junk filler that you would need to delete every time you actually want to write a new app, so we probably shouldn't put it in

danrabbit commented on data/templates/app/src/MainWindow.vala:
```diff
@@ -0,0 +1,93 @@
+        main_layout.attach (headerbar, 0, 0);
+        main_layout.attach (content, 0, 1);
+
+        var window_handle = new Hdy.WindowHandle ();
+        window_handle.add (main_layout);
```
It's probably not necessary to add a window handle for the entire window since you already have a headerbar which includes a windowhandle

meisenzahl replied:
With the `Hdy.WindowHandle` you can drag the application everywhere and not only on the headerbar. Isn't that something you want? 😅️

danrabbit replied:
Not necessarily. It's probably better to leave that to developers to decide if it makes sense for their app

danrabbit reviewed (changes requested):
This is definitely really interesting and speeds things up a lot to create a new app. I think in a lot of ways we should probably do less though, especially in places where you need to now go delete a bunch of placeholder stuff.

For the icon files, it would be better to have them named like `24.svg`, `32.svg` etc and use meson to rename them. In fact there's a lot of data files that could have simpler names and use `rename` in meson so that they would be easier to identify in the sidebar instead of having to use RDNN there

In the dialog, it probably would be better to use `Granite.ValidatedEntry` to validate inputs as you go instead of waiting to validate afterwards and throwing errors

meisenzahl commented on data/templates/app/data/com.github.{{ github_username }}.{{ github_repository }}.appdata.xml.in:
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This file can not be renamed with meson. Otherwise ninja will generate an error:

```
msgfmt: cannot locate ITS rules for ../data/appdata.xml.in
```

danrabbit replied:
You could probably make it significantly shorter still though and use `{{ github_repository }}.appdata.xml.in`

meisenzahl commented:
I added two more fields to the form for summary and description. I suggest that we split the form thematically into several pages with a following PR like in Onboarding. There I would also implement the logic to validate the input.

![Bildschirmfoto von 2021-04-24 08 32 52](https://user-images.githubusercontent.com/10796736/115949784-b534ea80-a4d7-11eb-8594-30be4235233d.png)

With another PR I would like to implement the logic to select a license.

@danrabbit what do you think of the approach?

danrabbit commented on data/templates/app/src/Application.vala:
```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright {{ current_year }} {{ your_name }} (https://github.com/{{ github_username }})
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program; if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA
+ *
+ * Authored by: {{ your_name }} <{{ your_email }}>
+ */
```
If we don't have the ability to choose the license, maybe we should leave this license header out for now. I think @tintou wanted to propose we switch to the SPDX license identfier anyways so it would be more like:

```
/*
 * SPDX-License-Identifier: LGPL-3.0-or-later
 * SPDX-FileCopyrightText: 2021 {{ your_name }}
 */
```

danrabbit commented on data/templates/app/src/MainWindow.vala:
```diff
@@ -0,0 +1,76 @@
+
+        var headerbar = new Hdy.HeaderBar () {
+            show_close_button = true,
+            title = "{{ app_name }}",
```
It looks like this doesn't set the title for the window itself, just in the headerbar. So, for example in multitasking view this window shows as `com.github.foo.bar` instead of the name of the app

danrabbit commented on src/Widgets/NewAppDialog.vala:
```diff
@@ -0,0 +1,268 @@
+            margin_top = 14
+        };
+
+        var form_grid = new Gtk.Grid ();
+        form_grid.margin_start = form_grid.margin_end = 12;
+        form_grid.orientation = Gtk.Orientation.VERTICAL;
+        form_grid.row_spacing = 3;
+        form_grid.valign = Gtk.Align.CENTER;
+        form_grid.vexpand = true;
```
Code style :)
```suggestion
        var form_grid = new Gtk.Grid () {
            orientation = Gtk.Orientation.VERTICAL,
            row_spacing = 3,
            valign = Gtk.Align.CENTER,
            vexpand = true
        };
        form_grid.margin_start = form_grid.margin_end = 12;
```

danrabbit commented on src/Widgets/NewAppDialog.vala:
```diff
@@ -0,0 +1,268 @@
+        form_grid.add (link_button);
+        form_grid.show_all ();
+
+        deletable = false;
+        modal = true;
+        resizable= false;
+        width_request = 560;
+        window_position = Gtk.WindowPosition.CENTER_ON_PARENT;
```
Since this is a Granite.Dialog, these properties are already set
```suggestion
        modal = true;
        resizable= false;
        width_request = 560;
```

danrabbit commented on src/Widgets/NewAppDialog.vala:
```diff
@@ -0,0 +1,268 @@
+        get_content_area ().add (form_grid);
+
+        var cancel_button = add_button (_("Cancel"), Gtk.ResponseType.CANCEL);
+        cancel_button.margin_bottom = 6;
+        cancel_button.margin_top = 14;
+
+        create_button = (Gtk.Button) add_button (_("Create App"), Gtk.ResponseType.OK);
+        create_button.margin = 6;
+        create_button.margin_start = 0;
+        create_button.margin_top = 14;
```
Granite.Dialog already has margins set for the action area
```suggestion

        create_button = (Gtk.Button) add_button (_("Create App"), Gtk.ResponseType.OK);
```

danrabbit reviewed (changes requested):
Much better! A few more comments and suggestions

danrabbit commented:
Cass and I talked a bit about this on a call this morning, so I just wanted to leave a record here:

Something that's a bit of a concern is that while it's cool to do a bunch of automation, it isn't necessarily conducive to learning on a new platform. I've seen quite a few folks have this experience in XCode, Android Studio, Builder, etc where using a template to get a fully working base app has a wow factor, but then the user is lost about what to do next and they don't actually understand how their app works.

I wonder if it would be better to have a much much more basic starting point and then have several template and/or snippets functions on top of that so that users still have some kind of process of assembling their app and they can understand what each part does. Theoretically creating a new app is something you don't do very often, so it's probably okay for there to be a little bit of friction here.

Already there's an idea of having a larger wizard with many more entries, but it might be a better experience to instead get a user to some simple digestible code that they can start hacking on and then have additional separate functions for adding more features like gsettings, etc. At the least, things like adding a new GitHub actions workflow, editorconfig, README, etc I think should be their own separate functions that users opt-in to intentionally as they want them.

danrabbit commented on src/Widgets/NewAppDialog.vala:
```diff
@@ -0,0 +1,263 @@
+
+        var your_name_label = new Granite.HeaderLabel (_("Your Name"));
+
+        your_name_entry = new Granite.ValidatedEntry () {
```
Can we pre-populate this from `GLib.Environment.get_real_name ()`?

danrabbit commented on src/Widgets/NewAppDialog.vala:
```diff
@@ -0,0 +1,263 @@
+            hexpand = true
+        };
+
+        var your_github_label = new Granite.HeaderLabel (_("Your GitHub Account"));
```
This should probably be `Username` so that it's clear we're not asking for a URL or something else

danrabbit commented on src/Widgets/NewAppDialog.vala:
```diff
@@ -0,0 +1,263 @@
+
+        var your_github_label = new Granite.HeaderLabel (_("Your GitHub Account"));
+
+        your_github_entry = new Granite.ValidatedEntry () {
```
Maybe this could be pre-populated from `GLib.Environment.get_user_name ()`? Might not always be correct, but probably a good guess

hanaral commented:
Hey, I wanted to add a couple of ideas for the initial 'New App' workflow:
- As Dan mentioned, the setup isn't actually this direct in XCode; I think it would be better to just have a button to 'Import App Template' and then have the obvious default template as the pre-installed option.
- Another thing to do with templates; how about giving the user a couple of checkboxes to control the amount of the template that is actually inherited? That way the problem of some developers only using X features and having to delete some of the stuff is solved (like not setting up the actual app code itself, but just all the other files that no-one wants to do like meson or appdata)

meisenzahl commented:
My idea was to use a following PR to group the entries on individual pages similar to the order in the developer documentation. So you could also link to the entry in the documentation where the steps are explained on the respective page.

I'm open to what features the template should contain. @danrabbit I would appreciate a list of what the template should and should not include. I will be happy to clean up here then 😁️

pablojimpas commented:
This is huge for new developers that are trying the platform. I like the idea of keeping it simple by default so it isn't overwhelming and newbies can get their head around what the template it's doing behind the scenes. 

At the same time I will like to see this concept extended to more powerful templates so it's easier to quick start a project and get early prototypes working in a couple of minutes.

I leave some prior art here for reference:

https://github.com/kjlaw89/archetype
https://github.com/microsoft/windowsTemplateStudio
https://github.com/microsoft/WebTemplateStudio

donadigo reviewed (changes requested):
Some comments for you.

danrabbit commented:
@jeremypw My inclination is that there shouldn't be a single feature to create an entire app as this kind of dumps you into an overwhelming amount of code all at once and the case of creating a whole brand new app is a rare case that's kind of weird to optimize for.

It might be more helpful to have a handful of separate features for creating each of the following:
* Vala class file
* meson build file
* AppData file
* Desktop file
* Flatpak Manifest
* LICENSE

This would still be really helpful to cut down on boilerplate, it would avoid creating a ton of files all at once that are hard to understand, and it would be more generally useful for projects that in varying states. I don't think it's an anti-feature to require some kind of documentation to create a brand new app from scratch.

I think combining this with some kind of snippets for each file type would be really powerful. For example, in an appdata file you'd have a snippet for a new release tag or in a Flatpak manifest you could have a snippet for a new module.

Marukesu commented:
> It might be more helpful to have a handful of separate features for creating each of the following

Isn't this the case of the TemplateManager?
I mean, the api seems pretty extensible, so they can be used in both cases.

jeremypw commented:
@meisenzahl From Danielle's comments it looks like the way forward would be to break this PR up into smaller diffs starting perhaps with the ability to choose one or two basic templates (e.g. stand-alone Vala class file, root meson.build file).  You would still need a dialog to gather information such as author name, email address, license, class name, project name, depending on the nature of the module.   This could then be iterated on to add further modules and refinements such as updating the meson.build file when another class is added.

Combining these with snippets (for which another PR is currently in progress) would reduce the donkey work without deducing the need for understanding.

meisenzahl commented:
I'm closing this because the desired solutions are likely to be solved with snippets and the `TemplateManager`.
